# Практическая работа 12: Исследование работы LLM в контексте работы с собственными числами матрицы

Ваше ФИО:

Вавилина Екатерина Андреевна

Что хотим сегодня изучить:
1. Реализации методов поиска собственных значений и собственных векторов на Python
2. Метод главных компонент, который много где используется сам по себе и основан на работе с собственными значениями матриц
3. Метод главных компонент для оптимизации (уменьшения размеров) модели LLM
4. Общение с AI-ассистентами (готовыми LLM) для исследования (изучения) новой предметной области (пункт 3 из списка ниже)
5. Научиться писать промпты к LLM так, чтобы получать адекватные и неадекватные результаты (галлюцинации)
6. Находить и устранять галлюцинации LLM в программировании и исследовании, когда предметная область Вам уже знакома (пункт 2 из списка ниже)
7. Научиться оптимизировать модели LLM и проверять их точность (пункт 1 из списка ниже)

В контексте использования large language model (LLM) термин "исследование" обычно относится к процессу изучения, разработки или применения LLM для достижения научных, технических или академических целей. Он может иметь несколько значений в зависимости от конкретного контекста:

1. **Разработка модели**: Исследование включает в себя создание или совершенствование LLM, включая проектирование архитектуры, оптимизацию алгоритмов обучения или повышение производительности (например, точности, результативности или обобщения). Этим часто занимаются исследователи ИИ или инженеры.

2. **Изучение возможностей**: Исследование может означать изучение того, что могут делать LLM, например, тестирование их способности решать конкретные задачи (например, понимать естественный язык, рассуждать или генерировать код) или оценку их ограничений и предубеждений.

3. **Применение к предметной области**: Использование LLMS для поддержки исследований в других областях, таких как анализ научной литературы, генерация гипотез или автоматизация обработки данных в таких дисциплинах, как медицина, физика или социальные науки.

4. **Влияние на этику и общество**: Изучение последствий внедрения LLM, включая справедливость, безопасность, риски дезинформации или экологические издержки, связанные с обучением и выводами.

Например, исследователь может использовать LLM для анализа обширных наборов данных для исследования (прикладной работы) или для экспериментов с тонкой настройкой, чтобы улучшить производительность при выполнении узкоспециализированной задачи (разработка). Термин является широким, но обычно подразумевает систематическое изучение или экспериментирование с участием LLM.

### **План работы (разделы ноутбука)**
1. Выбрать один или несколько AI-ассистентов из списка ниже и перечислить их ниже.
2. Написать промпт (желательно один, чтобы можно было корректнее сравнить результат) на любой языке, на котором Вы ожидаете получить лучший результат, в результате которого выбрать модель с весами LLM для скачивания. Далее будем работать с этой моделью как с матрицей. Сравните ответы полученные от разных LLM в таблице.
3. После этого с помощью того же или нового набора AI-ассистентов, получите сниппет для скачивания и начала работы с моделью. Вставьте сниппеты в соответствующие ячейки. Запустите ячейки и сравните полученные результаты, не внося дополнительных изменений в код, если он не работает.
4. Если код после запуска не работает, получите у соответствующего AI-ассистента рекомендации по исправлению кода. Вставьте исправленный код в новую соответствующую ячейку. Запустите код и в случае, если код снова не работает, получите новые исправления. Продолжайте либо, пока код не заработает, либо пока AI-ассистент не начнёт слишком сильно галлюцинировать. Протоколируйте все диалоги в соответствующие ячейки. Проведите анализ полученных результатов в текущем разделе.
5. Напишите собственные реализации методов нахождения собственных чисел и собственных векторов матриц, используя вызовы готовых реализаций в стандартных библиотеках или подробные описания.
6. Заставьте AI-агентов написать сниппеты для тех же методов. Сравните на любой матрце.
7. Исследуйте метод главных компонент для оптимизации размера выбранной модели при помощи собственных значений и собственных векторов. Сравните результаты работы модели до и после оптимизации.
8. Заключение.

---

## Раздел 1. Выбрать один или несколько AI-ассистентов из списка ниже и перечислить их ниже.

Список:

- [ChatGPT (OpenAI)](https://chatgpt.com/)
- [Claude (Anthropic)](https://claude.online/chat)
- [Gemini (Google)](https://gemini.google.com/app)
- [Grok (xAI)](https://grok.com/)
- [Perplexity](https://www.perplexity.ai/)
- [DeepSeek](https://chat.deepseek.com/)
- [Qwen](https://chat.qwen.ai/)
- [Llama (Meta AI)](https://huggingface.co/chat/) c учетной записью от Hugging Face
- [Mistral](https://mistral.ai/)
- [OpenAssistant](https://open-assistant.io/chat/)
- [GigaChat (Sber)](https://giga.chat/#chat)

- [Microsoft Copilot](https://copilot.microsoft.com/chats/)
- [Tabnine](https://www.tabnine.com/pricing/) - платно, но есть триал
- [Amazon Q Developer](https://aws.amazon.com/ru/q/developer/)
- [GitHub Copilot](https://github.com/copilot)

Вы также можете использовать модели, которые не попали в список, но обязательно указывайте их источник, название и прочие аттрибуты.

DeepSeek (deepThink R1 версия), ChatGPT (4O версия)

---

## Раздел 2. Написать промпт, отправить AI-ассистентам, сравнить результаты
2.1. Написать промпт (желательно один, чтобы можно было корректнее сравнить результат) на любой языке, на котором Вы ожидаете получить лучший результат, в результате которого **выбрать** модель с весами LLM для скачивания (желательно выбирать небольшие модели, чтобы скачивать было не много и матрицы были не слишком большими). Далее будем работать с этой моделью как с матрицей.

2.2. Отправить промпт AI-ассистентам, выбранным в разделе 1.

2.3. Сравните ответы полученные от разных LLM в таблице.

2.4. Выбрать модель для скачивания

### 2.1 Использованный промпт:

List 3–5 open-source LLMs (≤2 B params) for matrix/PCA work. For each, give:
- Name, params, download size, license, URL
- One-sentence rationale

Finally, recommend the single best model.

### 2.3. Сравнение результатов

| **Ассистент** | **Ответ**                                                                                                                                                                                                                                                                                  | **Ваш комментарий**                                                                   |
|---|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|---------------------------------------------------------------------------------------|
|*DeepSeek (deepThink R1 версия)*| *Microsoft Phi-1.5 is the top choice due to its specialized training on mathematical and coding datasets, MIT license for flexibility, and strong performance in analytical tasks like PCA. Its balance of size (1.3B) and domain relevance makes it uniquely suited for matrix operations.* | *На выбор были: 1. Microsoft Phi-1.5, 2. Pythia-1.4B, 3. TinyLlama-1.1B, 4. Bloom-560M* |
|*ChatGPT (4O версия)*| *For tasks involving matrix operations and PCA, GPT-2 is recommended due to its balance between parameter size and performance. It is lightweight, efficient, and capable of assisting in generating code for such tasks.* | *На выбор были: 1. GPT-2, 2. GPT-J 6B, 3. OpenLLaMA 3B*                        |

### 2.4. Выбранная модель

Попробуем взять GPT-2
Если в процессе пойдет что-то не так - возьмем Microsoft Phi-1.5 (и почему же гпт рекомендует гпт)

---

## Раздел 3. Скачивание модели с помощью сниппетов от AI-ассистентов
С помощью того же или нового набора AI-ассистентов, получите сниппет для скачивания и начала работы с моделью. Вставьте сниппеты в соответствующие ячейки. Запустите ячейки и сравните полученные результаты, не внося дополнительных изменений в код, если он не работает.

3.1. Список используемых AI

3.2. Сниппеты для скачивания выбранной модели

3.3. Запуск сниппетов

3.4. Анализ результатов

### 3.1. Список используемых AI

DeepSeek (deepThink R1 версия), ChatGPT (4-mini-high) <- поменяла модель на ту, которая лучше дружит с кодированием

Промпт:
Generate a Python code snippet that demonstrates how to download a GPT-2  model, load it, and use it to process input text, providing an example of how to use the model for inference.




### 3.2. Сниппеты для скачивания выбранной модели

Модель 1: DeepSeek

In [2]:
# Import required libraries
from transformers import AutoTokenizer, AutoModelForCausalLM
import torch

# Download and load the model and tokenizer
model_name = "gpt2"  # You can use "gpt2-medium", "gpt2-large", or "gpt2-xl" for larger versions
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForCausalLM.from_pretrained(model_name)

# Set the model to evaluation mode
model.eval()

def generate_text(prompt, max_length=50, temperature=0.7, top_k=50):
    """
    Generate text using GPT-2 model
    Args:
        prompt (str): Input text to start generation
        max_length (int): Maximum length of generated text
        temperature (float): Controls randomness (lower = more deterministic)
        top_k (int): Top-k sampling parameter
    Returns:
        str: Generated text
    """
    # Tokenize input text
    inputs = tokenizer.encode(prompt, return_tensors="pt")

    # Generate text
    with torch.no_grad():
        outputs = model.generate(
            inputs,
            max_length=max_length,
            temperature=temperature,
            top_k=top_k,
            pad_token_id=tokenizer.eos_token_id,
            do_sample=True,
            num_return_sequences=1
        )

    # Decode and return generated text
    generated_text = tokenizer.decode(outputs[0], skip_special_tokens=True)
    return generated_text

# Example usage
if __name__ == "__main__":
    input_text = "In a world where artificial intelligence"
    generated = generate_text(input_text, max_length=100)
    print("Input text:", input_text)
    print("Generated text:")
    print(generated)

Input text: In a world where artificial intelligence
Generated text:
In a world where artificial intelligence is king, the challenge is to make sure that future generations of researchers don't have to rely on human intelligence to make decisions.

For example, in a new study, researchers at the University of California, Santa Cruz, have shown that the human brain can recognize language when a person speaks it.

The work, titled "A more nuanced understanding of how language and cognition evolve in a complex social network," was recently published in the journal Frontiers in Neuroscience


Модель 2: ChatGPT

In [1]:
from transformers import GPT2LMHeadModel, GPT2Tokenizer

# Load the pre-trained GPT-2 model and tokenizer
model_name = 'gpt2'  # You can also use other variants like 'gpt2-medium', 'gpt2-large', or 'gpt2-xl'
model = GPT2LMHeadModel.from_pretrained(model_name)
tokenizer = GPT2Tokenizer.from_pretrained(model_name)

# Ensure padding is correctly handled
tokenizer.pad_token = tokenizer.eos_token

# Function to generate text based on input
def generate_text(prompt, max_length=100):
    # Encode the input prompt to tensor
    inputs = tokenizer.encode(prompt, return_tensors='pt')

    # Generate text with the model
    outputs = model.generate(inputs, max_length=max_length, num_return_sequences=1, no_repeat_ngram_size=2)

    # Decode the generated text
    generated_text = tokenizer.decode(outputs[0], skip_special_tokens=True)

    return generated_text

# Example usage
input_text = "Once upon a time"
generated_text = generate_text(input_text)

print("Input Text:", input_text)
print("Generated Text:", generated_text)



/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json:   0%|          | 0.00/665 [00:00<?, ?B/s]

Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP download. For better performance, install the package with: `pip install huggingface_hub[hf_xet]` or `pip install hf_xet`


model.safetensors:   0%|          | 0.00/548M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/1.04M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
The attention mask is not set and cannot be inferred from input because pad token is same as eos token. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.


Input Text: Once upon a time
Generated Text: Once upon a time, the world was a place of great beauty and great danger. The world of the gods was the place where the great gods were born, and where they were to live.

The world that was created was not the same as the one that is now. It was an endless, endless world. And the Gods were not born of nothing. They were created of a single, single thing. That was why the universe was so beautiful. Because the cosmos was made of two


### 3.4. Анализ результатов

Начала с выполнения ячейки от Chat GPT (потому что она сгенерировалась заметно быстрее). DeepSeek долго думал, прежде чем написать код.
- Код Chat GPT получился более компактным и более человекочитаемым. На первый згляд он использует больший функционал "из коробки", не усложняя выполнение.
- DeepSeek добавил больше комментариев к коду (точнее один многострочный комментарий). Получившийся снипет так же сделан с поддержкой импортирования этого скрипта как модуля, что для данной заачи (и работы в Коллабе) выглядит чуть более усложненно.  
- Обе предложенные реализации поймаои варнинг, связанный с работой Коллаба (т.к. код при загрузке модели одинаковый) и токенами. Это не очень хорошо, но т.к. в промпте не было уточнений на этот счет - это не страшно.

---

## Раздел 4. Первые галлюцинации
Если код после запуска не работает, получите у соответствующего AI-ассистента рекомендации по исправлению кода. Вставьте исправленный код в новую соответствующую ячейку. Запустите код и в случае, если код снова не работает, получите новые исправления. Продолжайте либо, пока код не заработает, либо пока AI-ассистент не начнёт слишком сильно галлюцинировать. Протоколируйте все диалоги в соответствующие ячейки. Проведите анализ полученных результатов в текущем разделе.

В случае, если все модели отработали безупречно, вернитесь к этому разделу после промта по написанию методов рассчёта собственных значений матрицы (раздел 6) или оптимизации выбранной модели LLM при помощи методов поиска собственных значений и собственных векторов матрицы весов. (раздел 7)

#### Справка: Когда LLM начинают галлюцинировать?

**LLMs склонны к галлюцинациям во время написания кода, главным образом, в таких ситуациях:**

1. **Неоднозначные или недостаточно определенные запросы**  
   - Если запрос расплывчатый ("напишите алгоритм быстрой сортировки"), то LLM может начать гадать и выдумывать детали.
2. **Редкие или незнакомые библиотеки, API или фреймворки**  
   - Если Вы спрашиваете о малоизвестном инструменте или совершенно новой версии библиотеки, LLM может "придумать" методы, классы или шаблоны использования на основе шаблонов, которые он видел в других местах.
3. **Междоменные запросы**  
   - При объединении технологий (например, "использовать модели SQLAlchemy внутри конвейера TensorFlow") могут возникнуть галлюцинации с API, которых на самом деле не существует.
4. **Устаревшие или будущие знания**  
   - Если что-то изменилось после завершения обучения модели (например, новые возможности версии Python), она может придумать поведение, которое "ожидает", основываясь на тенденциях.
5. **Сложные многоэтапные задачи**  
   - В более длинных цепочках (например, при полной настройке приложения) накапливаются ошибки - пропущенный импорт, неправильные типы, воображаемые функции и т.д.
6. **Когда вас просят оптимизировать, упростить или сильно обобщить**  
   - "Напишите наиболее эффективную систему кэширования" — это может привести к созданию идеализированных методов, которые звучат правдоподобно, но не являются реальными.
7. **Слишком самоуверенные формулировки в запросах**  
   - Если вы спросите "Дайте мне точный вызов API для..." с полной уверенностью, модель может выдать очень уверенный, но совершенно неверный ответ.

**Как распознать галлюцинации, когда LLMs пишут код:**

- **Сверьтесь с официальной документацией.**  
  Всегда сверяйте названия функций, сигнатуры методов и использование библиотек с реальной документацией.

- **Запустите и протестируйте код на ранней стадии.**  
  Даже небольшой пример (модульный тест, выполнение скрипта) может сразу выявить недостоверные детали.

- **Используйте средства проверки типов и линтеры.**  
  Такие инструменты, как "mypy", "pylint" или "eslint", часто обнаруживают (*выдуманные*) ошибочные API, неправильный импорт и т.д.

- **Задавайте дополнительные вопросы.**  
  Если вы не уверены, вы можете спросить у LLM: "Вы уверены, что `foo.bar()` существует? Можете ли вы показать документацию?". Но в последних версиях LLM поголовно научились нагло врать, наставать и выдумывать ссылки.

- **Будьте подозрительны к слишком точным ответам.**  
  Реальный код часто требует обработки ошибок, нестандартных ситуаций, шагов по настройке. Галлюцинированный код часто выглядит "слишком чистым".

- **Делайте подсказки конкретными.**  
  Сужение задачи снижает вероятность того, что модель ошибется в своих предположениях.

Модель 1: Сhat GPT. Фиксим 7 раздел.

Пытаюсь понять, оптимизировал он что-то в 7 задании или нет. Новый код по промпту:
а ты уверен, что реализовал все требования правильно?

In [9]:
import torch
import numpy as np
from sklearn.decomposition import PCA
from transformers import GPT2LMHeadModel, GPT2Tokenizer
from torch.utils.data import DataLoader, Dataset
from torch.nn.functional import log_softmax
from math import exp

# Load the pre-trained GPT-2 model and tokenizer
model_name = 'gpt2'
model = GPT2LMHeadModel.from_pretrained(model_name)
tokenizer = GPT2Tokenizer.from_pretrained(model_name)

# Set the padding token to be the EOS token
tokenizer.pad_token = tokenizer.eos_token

# Simulate a small dataset (50 examples)
class TinyDataset(Dataset):
    def __init__(self, tokenizer, num_samples=50):
        self.tokenizer = tokenizer
        self.num_samples = num_samples
        self.examples = ["Once upon a time"] * num_samples  # Simplified for the example

    def __len__(self):
        return self.num_samples

    def __getitem__(self, idx):
        text = self.examples[idx]
        encoding = self.tokenizer(text, return_tensors='pt', padding=True, truncation=True, max_length=50)
        return encoding['input_ids'].squeeze(0), encoding['attention_mask'].squeeze(0)

# Create a small data loader for testing
dataset = TinyDataset(tokenizer)
dataloader = DataLoader(dataset, batch_size=5)

# Function to compute Perplexity
def compute_perplexity(model, dataloader):
    model.eval()
    total_loss = 0.0
    total_tokens = 0
    for batch in dataloader:
        input_ids, attention_mask = batch
        with torch.no_grad():
            outputs = model(input_ids, attention_mask=attention_mask, labels=input_ids)
        loss = outputs.loss
        total_loss += loss.item() * input_ids.size(0)
        total_tokens += input_ids.size(0) * input_ids.size(1)

    # Compute perplexity
    perplexity = exp(total_loss / total_tokens)
    return perplexity

# Function to apply PCA on the model's weights (in this case, the first transformer layer's weights)
def optimize_with_pca(model, n_components=0.9):
    pca = PCA(n_components=n_components)

    # Apply PCA to all attention layers (Q, K, V) for all transformer layers
    for layer in model.transformer.h:
        attention_weights = layer.attn.c_attn.weight.detach().cpu().numpy()
        reduced_weights = pca.fit_transform(attention_weights)
        optimized_weights = pca.inverse_transform(reduced_weights)
        layer.attn.c_attn.weight.data = torch.tensor(optimized_weights).to(model.device)

    return pca

# Evaluate the model's perplexity before optimization
initial_perplexity = compute_perplexity(model, dataloader)
print(f"Perplexity before PCA optimization: {initial_perplexity:.4f}")

# Perform PCA optimization on the attention layers
pca = optimize_with_pca(model, n_components=0.9)

# Evaluate the model's perplexity after optimization
optimized_perplexity = compute_perplexity(model, dataloader)
print(f"Perplexity after PCA optimization: {optimized_perplexity:.4f}")

# Print some details about PCA (eigenvalues and components)
print("PCA Eigenvalues: ", pca.explained_variance_ratio_)
print("PCA Explained Variance (Eigenvalues): ", pca.explained_variance_)
print("PCA Components (Eigenvectors): ", pca.components_)


Perplexity before PCA optimization: 2.4442
Perplexity after PCA optimization: 3.4768
PCA Eigenvalues:  [0.0075548  0.0070551  0.00690758 0.00618688 0.00590634 0.00544115
 0.00535193 0.00510829 0.00503018 0.0049473  0.00489557 0.00478335
 0.00474998 0.00471715 0.0046343  0.00462774 0.00456169 0.00454372
 0.00450765 0.00444946 0.00442181 0.00440601 0.00435659 0.00434763
 0.00432051 0.00429657 0.00425392 0.00418551 0.00418389 0.00416017
 0.00412964 0.00409472 0.00405999 0.0040362  0.00399529 0.00398328
 0.00396821 0.00394455 0.00391109 0.00388791 0.00386657 0.00385118
 0.00381033 0.00379369 0.00378804 0.00374493 0.00372207 0.00370461
 0.00368033 0.00366752 0.00365378 0.00362857 0.0036146  0.00359039
 0.00357427 0.00355583 0.00355033 0.00352452 0.00350748 0.00349868
 0.00349323 0.0034468  0.00344053 0.00339008 0.00338532 0.00335175
 0.00334094 0.00333604 0.00331531 0.00327029 0.00325851 0.00324414
 0.00322506 0.00321214 0.00319206 0.00318472 0.00315687 0.00313401
 0.00310303 0.0030923  0.0

Оптимизация стала хуже. Повторная генерация ничего не изменила. Убрала контекст и начала новый чат.

In [13]:
import torch
import numpy as np
from transformers import GPT2LMHeadModel, GPT2Tokenizer
from sklearn.decomposition import PCA
from sklearn.metrics import accuracy_score
from tqdm import tqdm

# Load the GPT-2 model and tokenizer
model_name = "gpt2"
model = GPT2LMHeadModel.from_pretrained(model_name)
tokenizer = GPT2Tokenizer.from_pretrained(model_name)

# Set the pad token to the eos token (end of sequence token)
tokenizer.pad_token = tokenizer.eos_token

# Example of a small dataset for testing
# You can replace this with a real validation set (e.g., Tiny BERTScore dataset)
dataset = ["This is a test sentence." for _ in range(50)]  # 50 examples

# Function to tokenize the input data and get model outputs
def get_model_outputs(texts):
    inputs = tokenizer(texts, return_tensors="pt", padding=True, truncation=True)
    outputs = model(**inputs, labels=inputs["input_ids"])
    return outputs.logits

# Calculate the accuracy of the model's predictions (dummy accuracy for illustration)
def compute_accuracy(logits, inputs):
    predictions = logits.argmax(dim=-1)
    correct = (predictions == inputs["input_ids"]).sum().item()
    total = inputs["input_ids"].numel()
    return correct / total

# PCA Optimization for weight reduction
def apply_pca_to_weights(layer_name, n_components=50):
    # Extract the embedding layer weights
    weights = model.transformer.wte.weight.data.cpu().numpy()  # Embedding weights
    original_shape = weights.shape

    # Apply PCA for dimensionality reduction
    pca = PCA(n_components=n_components)
    reduced_weights = pca.fit_transform(weights)
    reconstructed_weights = pca.inverse_transform(reduced_weights)

    # Replace the original weights with the reconstructed reduced weights
    model.transformer.wte.weight.data = torch.tensor(reconstructed_weights, dtype=torch.float32).to(model.device)

    print(f"Applied PCA to {layer_name} layer, reducing dimensions from {original_shape[1]} to {n_components}.")

# Test the model before optimization
def test_model_before_optimization():
    logits = get_model_outputs(dataset)
    accuracy = 0
    for i, text in enumerate(dataset):
        inputs = tokenizer(text, return_tensors="pt", padding=True, truncation=True)
        accuracy += compute_accuracy(logits[i], inputs)
    return accuracy / len(dataset)

# Test the model after optimization
def test_model_after_optimization():
    logits = get_model_outputs(dataset)
    accuracy = 0
    for i, text in enumerate(dataset):
        inputs = tokenizer(text, return_tensors="pt", padding=True, truncation=True)
        accuracy += compute_accuracy(logits[i], inputs)
    return accuracy / len(dataset)

# Apply PCA and test the optimization
def main():
    # Test accuracy before optimization
    accuracy_before = test_model_before_optimization()
    print(f"Accuracy before optimization: {accuracy_before * 100:.2f}%")

    # Apply PCA to reduce the weights in the embedding layer
    apply_pca_to_weights('embedding', n_components=50)

    # Test accuracy after optimization
    accuracy_after = test_model_after_optimization()
    print(f"Accuracy after optimization: {accuracy_after * 100:.2f}%")

    # Criterion for optimization
    improvement = accuracy_after - accuracy_before
    print(f"Improvement in accuracy: {improvement * 100:.2f}%")

if __name__ == "__main__":
    main()


Accuracy before optimization: 0.00%
Applied PCA to embedding layer, reducing dimensions from 768 to 50.
Accuracy after optimization: 0.00%
Improvement in accuracy: 0.00%


Пытаемся указать дополнительно на необходимость использовать TinyBERT

In [15]:
!pip install bert-score

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.1/61.1 kB 6.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 363.4/363.4 MB 4.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.8/13.8 MB 64.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.6/24.6 MB 37.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 883.7/883.7 kB 50.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 664.8/664.8 MB 2.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 211.5/211.5 MB 6.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.3/56.3 MB 14.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 127.9/127.9 MB 8.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 207.5/207.5 MB 6.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.1/21.1 MB 84.6 MB/s eta 0:00:00
  Attempting uninstall: nvidia-nvjitlink-cu12
    Found existing installation: nvidia-nvjitlink-cu12 12.5.82
    Uninstalling

In [5]:
import torch
import numpy as np
from transformers import GPT2LMHeadModel, GPT2Tokenizer
from sklearn.decomposition import PCA
from bert_score import score
import transformers
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA as sklearnPCA
from huggingface_hub import login

# Авторизация на Hugging Face (если требуется доступ к приватным моделям)
# login(token='your_token_here')  # Используйте ваш токен, если нужно

# Загрузка модели GPT-2 и токенизатора
model_name = "gpt2"
model = GPT2LMHeadModel.from_pretrained(model_name)
tokenizer = GPT2Tokenizer.from_pretrained(model_name)

# Устанавливаем pad_token на eos_token, если он не задан
tokenizer.pad_token = tokenizer.eos_token

# Загрузка стандартной модели BERT для BERTScore
bert_model_name = 'bert-base-uncased'  # Используем стандартную модель BERT
bert_model = transformers.BertModel.from_pretrained(bert_model_name)
bert_tokenizer = transformers.BertTokenizer.from_pretrained(bert_model_name)

# Пример небольшого набора данных для тестирования
dataset = ["This is a test sentence." for _ in range(50)]  # 50 примеров
reference = ["This is a test sentence." for _ in range(50)]  # Референсный текст для оценки

# Функция для получения выводов модели
def get_model_outputs(texts):
    inputs = tokenizer(texts, return_tensors="pt", padding=True, truncation=True)
    outputs = model(**inputs, labels=inputs["input_ids"])
    return outputs.logits

# PCA для оптимизации весов
def apply_pca_to_weights(layer_name, n_components=50):
    # Извлекаем веса эмбеддинга
    weights = model.transformer.wte.weight.data.cpu().numpy()  # Веса эмбеддинга
    original_shape = weights.shape

    # Применяем PCA для уменьшения размерности
    pca = sklearnPCA(n_components=n_components)
    reduced_weights = pca.fit_transform(weights)
    eigenvalues = pca.explained_variance_
    eigenvectors = pca.components_

    # Пересчитываем восстановленные веса после PCA
    reconstructed_weights = pca.inverse_transform(reduced_weights)

    # Заменяем оригинальные веса на уменьшенные
    model.transformer.wte.weight.data = torch.tensor(reconstructed_weights, dtype=torch.float32).to(model.device)

    print(f"Применен PCA к слою {layer_name}, уменьшив размерность с {original_shape[1]} до {n_components}.")
    print(f"Собственные значения: {eigenvalues}")
    print(f"Собственные векторы: {eigenvectors}")

# Оценка качества модели с использованием BERTScore
def evaluate_with_bertscore(model, dataset, reference):
    generated_texts = []
    for text in dataset:
        inputs = tokenizer(text, return_tensors="pt", padding=True, truncation=True)
        outputs = model.generate(inputs["input_ids"], max_length=50, num_return_sequences=1,
                                 pad_token_id=tokenizer.eos_token_id, attention_mask=inputs["attention_mask"])
        generated_text = tokenizer.decode(outputs[0], skip_special_tokens=True)
        generated_texts.append(generated_text)

    # Вычисляем BERTScore с использованием стандартной модели BERT
    P, R, F1 = score(generated_texts, reference, model_type=bert_model_name)
    return torch.mean(F1)  # Применяем torch.mean для корректного вычисления среднего значения

# Тестирование модели до оптимизации
def test_model_before_optimization():
    score_before = evaluate_with_bertscore(model, dataset, reference)
    return score_before

# Тестирование модели после оптимизации
def test_model_after_optimization():
    score_after = evaluate_with_bertscore(model, dataset, reference)
    return score_after

# Основная функция
def main():
    # Оценка до оптимизации
    score_before = test_model_before_optimization()
    print(f"BERTScore до оптимизации: {score_before:.4f}")

    # Применение PCA для уменьшения весов эмбеддинга
    apply_pca_to_weights('embedding', n_components=50)

    # Оценка после оптимизации
    score_after = test_model_after_optimization()
    print(f"BERTScore после оптимизации: {score_after:.4f}")

    # Критерий оптимизации
    improvement = score_after - score_before
    print(f"Улучшение в BERTScore: {improvement:.4f}")

if __name__ == "__main__":
    main()

BERTScore до оптимизации: 0.6993
Применен PCA к слою embedding, уменьшив размерность с 768 до 50.
Собственные значения: [0.21092454 0.09588385 0.08349932 0.08004653 0.07619064 0.07123205
 0.06563716 0.06398233 0.0608951  0.05769768 0.05716538 0.05548853
 0.05486828 0.05366705 0.05176363 0.05061083 0.04997892 0.04875418
 0.04761207 0.04601467 0.04531009 0.04459514 0.04323281 0.04301295
 0.04260088 0.0415249  0.041143   0.04046989 0.04004145 0.03978171
 0.03909929 0.03864159 0.03806955 0.03756968 0.0373752  0.03702661
 0.0369066  0.03678094 0.03623966 0.03618449 0.03561984 0.03527708
 0.03497235 0.03470382 0.03437415 0.03423247 0.03397272 0.03388649
 0.03356962 0.03338457]
Собственные векторы: [[ 1.8414611e-02  2.4681708e-02 -2.1835236e-02 ... -1.7881518e-02
  -1.0493988e-02 -4.6813328e-02]
 [ 1.9506289e-02  9.6143242e-03 -1.5648080e-02 ...  2.2144305e-02
  -1.3101786e-02 -1.0279728e-02]
 [-2.0713249e-02 -1.5950248e-02  7.2769038e-02 ...  3.8263430e-03
   2.1643827e-05 -1.9767084e-03]
 .

Не помогло, оптимизация делает только хуже (

---

## Раздел 5. Набросаем код методов самостоятельно из готовой документации

Реализуйте расчёт собственных значений матриц весов любого слоя при помощи известных Вам методов самостоятельно, используя стандартные библиотеки Python.

Выполните это задание самостоятельно, а не при помощи ассистентов (**любых**, включая GitHub Copilot autocomplite, его можно отключить в настройках справа внизу, Google Colab Gemini тоже можнот отключить в настройках "Инструменты"-"ИИ-функции"-"Предлагать подсказки ИИ для дополнения кода"), чтобы можно было корректно сравнить результаты стандартных библиотек и сниппетов, которые предложат Вам ассистенты далее.


| Метод | Функция Python | Библиотека | Ссылка на документацию |
|---|---|---|---|
|Power Method (Dominant Eigenvalue) | Manual (template provided) | Custom/Numpy | [Tutorial](https://bvanderlei.github.io/jupyter-guide-to-linear-algebra/Approximating_Eigenvalues.html#power-method) or [template](https://en.wikipedia.org/wiki/Power_iteration#Numerical_example) — implement with `numpy.dot` |
|Shifted Power Method | Manual + `numpy.linalg.solve` | Numpy | [Tutorial](https://bvanderlei.github.io/jupyter-guide-to-linear-algebra/Approximating_Eigenvalues.html) |
|QR Algorithm (Full Spectrum) | `scipy.linalg.eig` or `numpy.linalg.eig` | Scipy/Numpy | [scipy.linalg.eig](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.eig.html) / [numpy.linalg.eig](https://numpy.org/doc/stable/reference/generated/numpy.linalg.eig.html) |
|Full QR Decomposition (optional deeper understanding) | `numpy.linalg.qr` | Numpy | [numpy.linalg.qr](https://numpy.org/doc/stable/reference/generated/numpy.linalg.qr.html) |

In [6]:
import numpy as np

A = np.array([[4, 1], [2, 3]])

def qr_algorithm(A, num_iterations=1000):
    A_copy = A.copy()
    for _ in range(num_iterations):
        Q, R = np.linalg.qr(A_copy)
        A_copy = np.dot(R, Q)
    return np.diagonal(A_copy)


eig = qr_algorithm(A)
print(f"Eigenvalues: {eig}")


Eigenvalues: [5. 2.]


---

## Раздел 6. AI-ассистенты пишут код для расчёта собственных значений и векторов матриц
6.1. Напишите промпт для получения сниппетов методов расчёта собственных значений и векторов матриц.

6.2. Сниппеты для расчёта собственных значений и векторов матриц

6.3. Запуск сниппетов против любой выбранной Вами матрицы

6.4. Анализ результатов

### 6.1 Ваш промпт

Provide Python code snippets for calculating eigenvalues and eigenvectors of the matrix A: A = ([[4, 1], [2, 3]])
Please include the necessary imports and how to use each method with the given matrix A.

### 6.2. Сниппеты для расчёта собственных значений и векторов матриц

Модель 1: DeepSeek

In [4]:
import numpy as np

A = np.array([[4, 1],
              [2, 3]])

# Find eigenvalues via characteristic equation
trace = np.trace(A)
det = np.linalg.det(A)
eigenvalues = np.roots([1, -trace, det])

print("Eigenvalues:", eigenvalues)

# Find eigenvectors manually
for λ in eigenvalues:
    B = A - λ * np.eye(2)
    _, _, vh = np.linalg.svd(B)
    eigenvector = vh[-1]
    print(f"Eigenvector for {λ:.2f}: {eigenvector}")

Eigenvalues: [5. 2.]
Eigenvector for 5.00: [0.70710678 0.70710678]
Eigenvector for 2.00: [ 0.4472136  -0.89442719]


Модель 2: ChatGPT

In [6]:
# Import necessary libraries
import numpy as np

# Define the matrix A
A = np.array([[4, 1], [2, 3]])

# Calculate eigenvalues and eigenvectors
eigenvalues, eigenvectors = np.linalg.eig(A)

# Print the results
print("Eigenvalues:")
print(eigenvalues)

print("\nEigenvectors:")
print(eigenvectors)

Eigenvalues:
[5. 2.]

Eigenvectors:
[[ 0.70710678 -0.4472136 ]
 [ 0.70710678  0.89442719]]


### 6.3. Запуск сниппетов на любой выбранной матрице и численное сравнение с вашей реализацией

In [7]:
import numpy as np

A = np.array([[6, 2], [1, 5]])

# Мой код
def qr_algorithm(A, num_iterations=1000):
    A_copy = A.copy()
    for _ in range(num_iterations):
        Q, R = np.linalg.qr(A_copy)
        A_copy = np.dot(R, Q)
    return np.diagonal(A_copy)

# Код DeepSeek
def dipstick_method(A):
    trace = np.trace(A)
    det = np.linalg.det(A)
    eigenvalues = np.roots([1, -trace, det])

    eigenvectors = []
    for λ in eigenvalues:
        B = A - λ * np.eye(2)
        _, _, vh = np.linalg.svd(B)
        eigenvector = vh[-1]
        eigenvectors.append(eigenvector)

    return eigenvalues, eigenvectors

# Код Chat GPT
def gpt_method(A):
    eigenvalues, eigenvectors = np.linalg.eig(A)
    return eigenvalues, eigenvectors


print("Мой код - Собственные значения:")
eig_qr = qr_algorithm(A)
print(eig_qr)

print("\nКод дипсика - Собственные значения и векторы:")
eigenvalues_dip, eigenvectors_dip = dipstick_method(A)
print("Eigenvalues:", eigenvalues_dip)
print("Eigenvectors:", eigenvectors_dip)

print("\nКод GPT - Собственные значения и векторы:")
eigenvalues_gpt, eigenvectors_gpt = gpt_method(A)
print("Eigenvalues:", eigenvalues_gpt)
print("Eigenvectors:", eigenvectors_gpt)


Мой код - Собственные значения:
[7. 4.]

Код дипсика - Собственные значения и векторы:
Eigenvalues: [7. 4.]
Eigenvectors: [array([0.89442719, 0.4472136 ]), array([ 0.70710678, -0.70710678])]

Код GPT - Собственные значения и векторы:
Eigenvalues: [7. 4.]
Eigenvectors: [[ 0.89442719 -0.70710678]
 [ 0.4472136   0.70710678]]


### 6.4. Анализ полученных результатов

|Модель|Комментарий|
|---|---|
|Me, myself and I| Изобретаем велосипед, проверяем только собственные значения|
|DeepSeek| Знак для собственного вектора отличается от результата GPT, для векторов это нормально.|
|Chat GRT| Быстро, красиво, в одну строку, "из коробки". Т.к. код маленький и простой - не возникает проблем с ошибками компиляции.|

В итоге получились верные, работающий крограммы (с точностью до знака векторов, но это не является ошибкой). При этом DeepSeek выдал сразу несколько решений, что не требовалось в промпте. GPT при этом ограничилось одним решением, как и запрашивалось.

---

## Раздел 7. Оптимизация LLM при помощи метода главных компонент

### Справка: Метод главных компонент простыми словами с акцентом на собственные числа и собственные векторы

**Определение**

Метод главных компонентов (PCA - Principal Component Analysis) это подход к уменьшению размерности, такой что:
1. Находит новые оси (основные компоненты), где данные отличаются больше всего.
2. Проецирует данные на меньшее количество измерений, сохраняя при этом максимальную дисперсию.

**Как это работает с позиции собсвенных чисел и векторов:**
1. Берем матрицу данных `X` (например, включения токенов или скрытые слои LLM).
2. Вычисляем ковариационную матрицу:  
   $$
   C = \frac{1}{n} X^T X
   $$
3. Находим собственные числа и собственные векторы ковариационной матрицы`C`.
   - *Собственные векторы* = направления (главные компоненты)
   - *Собственные числа* = величина отклонения (дисперсии), зафиксированная в каждом направлении
4. Выберем только *топ-k собственных векторов* (наибольшие собственные числа) → получим сокращённое число измерений.

**Значение метода главных компонент для оптимизации размера LLM**

1. Сжатия слоев встраивания (embedding kayers) или скрытых представлений
- Пример: Вы сокращаете 768-мерные встраивания BERT до 128D, используя 128 лучших собственных векторов → огромная экономия памяти.
- Полезно для **дистилляции**, **квантования** или **работе на устройстве**.

2. Удалить избыточность в пространствах внимания (attention matrix)/фичах (Feature spaces)
- Заголовки внимания часто имеют "низкоуровневую структуру" (например, всего несколько доминирующих собственных мод).
- PCA может выявлять и "сокращать" или объединять заголовки, которые обладают сходными характеристиками.

3. Проанализировать репрезентативную способность
Построение собственного значения **спектра** скрытых состояний говорит вам:
- Насколько выразительная часть модели сосредоточена всего в нескольких направлениях.
- Является ли результат работы слоя **низкоуровневым** (сжимаемым).


### Визуальное представление
- **Плоская кривая собственных значений** → разнообразные, высокоуровневые представления (сложнее сжимать).
- **Крутой спад собственных значений** → большая часть информации сосредоточена в нескольких направлениях (легко сжимается с помощью PCA).

### Реальное приложение:
Google, Meta и HuggingFace использовали PCA и **низкоуровневую аппроксимацию** (например, усеченный SVD, линейные узкие места) для:
  - Сжатия моделей для мобильного развертывания.
  - Ускорения вывода.
  - Понять, на что модель "обращает внимание".



**PCA использует собственные значения/векторы ковариационной матрицы для определения доминирующих направлений в данных. В LLMs PCA выявляет избыточную структуру и обеспечивает сжатие путем проецирования вложений/выходных данных в подпространство меньшей размерности, что делает модели более быстрыми и компактными без существенной потери точности.**

7.1. Протокол запросов к выбранному (одному) ассистенту. Вы должны убедить его предоставить вам код оптимизации модели LLM (с которой Вы начинали работать в разделах 3-4), использующий метод главных компонентов с акцентом на использование методов расчёта собственных значений и собсвенных векторов. Также необходимо протестировать точность до и после оптимизации при помощи сокращённого набора данных (например, Tiny BERTScore (50 примеров)). Для оптимизации слой можно использовать любой, какой выберете или какой насоветует AI. Не обязательно сокращать все слои.

7.2. Полученный в результате код (обновляете в процессе работы над этим разделом, можно оставить только то, что получится в итоге)

7.3. Анализ трудностей и галлюцинаций на этом этапе. Как вы с этим справлялись (выявляли и исправляли).

7.4. Анализ полученной оптимизированной модели и применения метода

### 7.1. Протокол запросов к выбранному ассистенту.

Укажите, какой выбрали в итоге AI-ассистент. Приведите список промптов, которые вы пробовали.

Chat GPT (4o-mini-high)

**Промпт 1**:
Provide Python code for optimizing an LLM model using Principal Component Analysis (PCA), focusing on the use of eigenvalue and eigenvector computation methods. The code should:
1. Include PCA implementation for dimensionality reduction of the model's weights (you can choose the layer to optimize or suggest one based on your knowledge).
2. Apply eigenvalue and eigenvector computations during the optimization process.
3. Test the model's accuracy before and after optimization using a reduced dataset (e.g., Tiny BERTScore with 50 examples).
4. Include any necessary imports, setup, and details for testing accuracy.

Please suggest a layer to optimize if unsure.

**Промпт 2** (для исправления ошибки): ValueError: Asking to pad but the tokenizer does not have a padding token. Please select a token to use as `pad_token` `(tokenizer.pad_token = tokenizer.eos_token e.g.)` or add a new pad token via `tokenizer.add_special_tokens({'pad_token': '[PAD]'})`.

**Промпт 3**: Проанализируй вывод кода (он прикреплен выше). Ты точно добился оптимизации?
**Промтп 4**: Perplexity before PCA optimization: 2.4442, Perplexity after PCA optimization: 3.4768. Оптимизация не удалась, перегенерируй код.

### 7.2. Полученный код (в результате совместной работы и доработок):

In [8]:
import torch
import numpy as np
from sklearn.decomposition import PCA
from transformers import GPT2LMHeadModel, GPT2Tokenizer
from sklearn.metrics import accuracy_score
from torch.utils.data import DataLoader, Dataset
import random

# Load the pre-trained GPT-2 model and tokenizer
model_name = 'gpt2'
model = GPT2LMHeadModel.from_pretrained(model_name)
tokenizer = GPT2Tokenizer.from_pretrained(model_name)

# Set the padding token to be the EOS token
tokenizer.pad_token = tokenizer.eos_token

# Simulate a small dataset (50 examples)
class TinyDataset(Dataset):
    def __init__(self, tokenizer, num_samples=50):
        self.tokenizer = tokenizer
        self.num_samples = num_samples
        self.examples = ["Once upon a time"] * num_samples  # Simplified for the example

    def __len__(self):
        return self.num_samples

    def __getitem__(self, idx):
        text = self.examples[idx]
        encoding = self.tokenizer(text, return_tensors='pt', padding=True, truncation=True, max_length=50)
        return encoding['input_ids'].squeeze(0), encoding['attention_mask'].squeeze(0)

# Create a small data loader for testing
dataset = TinyDataset(tokenizer)
dataloader = DataLoader(dataset, batch_size=5)

# Function to evaluate model accuracy (simplified for illustration purposes)
def evaluate_model(model, dataloader):
    model.eval()
    predictions = []
    targets = []
    for batch in dataloader:
        input_ids, attention_mask = batch
        with torch.no_grad():
            outputs = model(input_ids, attention_mask=attention_mask, labels=input_ids)
        logits = outputs.logits
        preds = torch.argmax(logits, dim=-1)
        predictions.append(preds)
        targets.append(input_ids)

    # Flatten predictions and targets
    predictions = torch.cat(predictions).numpy()
    targets = torch.cat(targets).numpy()

    # Compute accuracy (simplified as token-wise comparison)
    accuracy = accuracy_score(targets.flatten(), predictions.flatten())
    return accuracy

# Function to apply PCA on the model's weights (in this case, the first transformer layer's weights)
def optimize_with_pca(model, layer_idx=0, n_components=0.9):
    # Extract the weights of the chosen layer (first transformer layer, hidden states)
    layer = model.transformer.h[layer_idx]

    # Get the weights of the attention layers (Q, K, V) and the linear layers
    attention_weights = layer.attn.c_attn.weight.detach().cpu().numpy()

    # Apply PCA on the attention weights
    pca = PCA(n_components=n_components)
    reduced_weights = pca.fit_transform(attention_weights)

    # Reconstruct the weights with reduced components
    optimized_weights = pca.inverse_transform(reduced_weights)

    # Set the optimized weights back to the model
    model.transformer.h[layer_idx].attn.c_attn.weight.data = torch.tensor(optimized_weights).to(model.device)

    # Return the PCA object for inspecting eigenvalues and eigenvectors
    return pca

# Evaluate the model's performance before optimization
initial_accuracy = evaluate_model(model, dataloader)
print(f"Accuracy before PCA optimization: {initial_accuracy:.4f}")

# Perform PCA optimization on the first transformer layer
pca = optimize_with_pca(model, layer_idx=0, n_components=0.9)

# Evaluate the model's performance after optimization
optimized_accuracy = evaluate_model(model, dataloader)
print(f"Accuracy after PCA optimization: {optimized_accuracy:.4f}")

# Print some details about PCA (eigenvalues and components)
print("PCA Eigenvalues: ", pca.explained_variance_ratio_)
print("PCA Explained Variance (Eigenvalues): ", pca.explained_variance_)
print("PCA Components (Eigenvectors): ", pca.components_)

`loss_type=None` was set in the config but it is unrecognised.Using the default loss: `ForCausalLMLoss`.


Accuracy before PCA optimization: 0.0000
Accuracy after PCA optimization: 0.0000
PCA Eigenvalues:  [0.02843004 0.02320209 0.02218907 0.01118033 0.00919784 0.00784012
 0.00700603 0.00694292 0.00678673 0.00674207 0.00664981 0.00647773
 0.00645821 0.00633618 0.0061688  0.00610498 0.00610272 0.00595685
 0.00594769 0.00588136 0.00585627 0.00580269 0.00571554 0.00567108
 0.00556388 0.00551927 0.00548839 0.00542392 0.00539786 0.00529561
 0.00528494 0.00525929 0.00521568 0.00521015 0.00507888 0.00501852
 0.0049875  0.00492477 0.00490224 0.00485763 0.00476582 0.00473946
 0.00466688 0.00462832 0.00454606 0.00450011 0.00446194 0.00435948
 0.00431916 0.00428984 0.00421963 0.00420975 0.00418572 0.00410201
 0.00401405 0.00397065 0.00389099 0.00386044 0.00383655 0.00375145
 0.00371842 0.00370128 0.0036226  0.00356542 0.00354041 0.00347301
 0.00335772 0.00334709 0.00327665 0.00325667 0.00321002 0.00320394
 0.00318656 0.00312843 0.00309987 0.00304752 0.00302162 0.0030065
 0.00298803 0.00296248 0.002913

### 7.3. Анализ трудностей и галлюцинаций на этом этапе. Как вы с этим справлялись (выявляли и исправляли).

Вообще я начинала с DeepSeek, но сервер занят. Это считается галлюцинацией?

Сгенерированный гпт первый код ошибся с токенайзером для GPT-2. Похоже, что из-за особенностей модели, пэддинг токен отстутсвует. Возможно, GPT поетрял контекст или не изпользовал предыдущий написанный код и ошибся в генерации. После передачи прямого текста ошибки гпт исправил код на работоспособный, с использованием eos_token.

Попробовала уточнить. есть ли оптимизация, т.к. судя по выводу оптимизации не было достигнуто в принципе. Дальнейшие мучения в разделе 4.



### 7.4. Анализ полученной оптимизированной модели и применения метода

Оптимизировался только первый слой, однако судя по выводу оптимизации не получилось (я пыталась, но видимо или в промпте ошибка, или я не совсем понимаю, куда его надо подтолкнуть, что бы код показывал оптимизацию). Финальным кодом можно считать тот, котрый находится в ячеке раздела 7, т.к. поптки исправить галлюцинации привели или к улучшению, или не изминили ничего (коды в разделе 4).

---

Что хочется добавить к сегодняшнему занятию:

1. [Код](https://github.com/tongjingqi/MathTrap) со ссылкой на статью про математические ловушки для LLM

2. [Заметка](https://habr.com/ru/articles/904754/) о том, какие навыки *кажется следует* подтягивать программистам в современной индустрии

3. Работая с различными AI-ассистентами, по крайней мере в Университете, будьте готовы к тому, что Вы не сможете получить максимум баллов за выполненные с их помощью работы, если только в задании не требуется обратное. Я провела опрос среди преподавателей нашего факультета (участие приняли 30 человек). 50% преподавателей высказались о том, что не поставят максимальный балл за работу выполненную с помощью GPT и аналогов, предлагая урезать баллы минимум в половину. Ещё 30% процентов поставят 0 баллов за такую работу и только лишь 20% преподавателей согласны принять работу выполненную при помощи ИИ-агентов и оценивать как выполненную самостоятельно.

## Итог

Расскажите о Вашем сегодняшнем опыте: если это привычный способ изучения новой области, прототипирования, то на сколько удалось погрузиться в тему собственных чисел и оптимизации LLM. Если вообще все Вам показалось новым, включая общение с LLM, то расскажите, как Вы справлись с написанием промптов и кода? Если писали пропты на языке отличном от родного, то на сколько трудно это Вам показалось и какие подходы использовали (например, машинный перевод?)
- Погружение в собственные числа и LLM прошло успешно, но я предпочитаю в таких случаях скорее искать через LLM объяснения алгоритмов и какую-то иинформацию со ссылками на источники, а не готовый код. Поэтому написание большого кода доставило достаточно много неприятностей.
- Мне особенно понравился "сервер занят" от DeepSeek. Про промпты: на английском в целом нормально получилось, но т.к. уточнения для модели я писала на русском - пояснения модели выдавали на русском.

Как Вы сами оцениваете свой прогресс на сегодняшнем занятии?
- Сегодня получились промпты получше. Мой навык их написания повысился.

Удалось ли Вам изучить всё то, что перечисленно в самом первом списке в ноутбуке? Что вызвало наибольшее затруднение?
- Удалос. Трудности: получение от AI действительно рабочего результата с учетом достаточно большого количества мелких ошибок, устранение галлюцинаций в коде; оптимизация моделей доставила немного боли.